prov/sm2/src/sm2_coordination.c

```diff
@@ -0,0 +1,631 @@
+#define SM2_STARTUP_MAX_TRIES	 1000
+
+static void sm2_file_attempt_shrink(struct sm2_mmap *map);
+int sm2_entry_lookup(const char *name, struct sm2_mmap *map);
+void *sm2_file_extend_for_region(struct sm2_mmap *map, int last_valid_entry);
+
+static inline bool sm2_mapping_long_enough_check(struct sm2_mmap *map, int item)
+{
+	ptrdiff_t entry_offset =
+		(char *) sm2_mmap_ep_region(map, item + 1) - map->base;
+	return entry_offset <= map->size;
+}
+
+static inline bool pid_lives(int pid)
+{
+	int err = kill(pid, 0);
+	return err == 0;
+}
+
+/*
+ * Take an open file, and mmap its contents
+ */
+void *sm2_mmap_map(int fd, struct sm2_mmap *map)
+{
+	struct stat st;
+
+	if (OFI_UNLIKELY(fstat(fd, &st))) {
+		FI_WARN(&sm2_prov, FI_LOG_EP_CTRL, "Failed fstat");
+		goto out;
+	}
+
+	map->base =
+		mmap(0, st.st_size, PROT_READ | PROT_WRITE, MAP_SHARED, fd, 0);
+	if (OFI_UNLIKELY(map->base == MAP_FAILED)) {
+		FI_WARN(&sm2_prov, FI_LOG_EP_CTRL,
+			"Failed mmap, st.st_size=%ld\n", st.st_size);
+		goto out;
+	}
+	map->fd = fd;
+	map->size = st.st_size;
+	return map->base;
+out:
+	map->base = NULL;
+	return NULL;
+}
+
+/*
+ * Take a sm2_mmap, and re-map if necessary, ensuring size of at_least
+ *
+ * If the size of the current map or the size of the file are not sufficient
+ * to address "at_least" bytes, then the file will be truncated() (extended)
+ * to the required size and the memory munmap()ed and re mmap()'ed
+ */
+void *sm2_mmap_remap(struct sm2_mmap *map, size_t at_least)
+{
+	struct stat st;
+
+	assert(at_least > 0);
+	if (map->size >= at_least)
+		return map->base;
+
+	if (OFI_UNLIKELY(fstat(map->fd, &st))) {
+		FI_WARN(&sm2_prov, FI_LOG_EP_CTRL,
+			"Failed fstat of sm2_mmaps file: %s, at_least: %zu\n",
+			strerror(errno), at_least);
+		return map->base;
+	}
+
+	if (st.st_size < at_least) {
+		if (OFI_UNLIKELY(ftruncate(map->fd, at_least))) {
+			FI_WARN(&sm2_prov, FI_LOG_EP_CTRL,
+				"Failed ftruncate of sm2_mmaps file: %s, "
+				"at_least: %zu\n",
+				strerror(errno), at_least);
+			return map->base;
+		}
+	} else if (st.st_size >= map->size) {
+		/* file has been extended by another process since
+		   last we checked.  Re-map the entire file. */
+		at_least = st.st_size;
+	} else {
+		FI_WARN(&sm2_prov, FI_LOG_AV,
+			"Shm file has shrunk, re-mapping!\n");
+		at_least = st.st_size;
+	}
+
+	if (map->size != at_least) {
+		if (OFI_UNLIKELY(munmap(map->base, map->size))) {
+			FI_WARN(&sm2_prov, FI_LOG_EP_CTRL,
+				"Failed unmap of sm2_mmaps file: %s, at_least: "
+				"%zu\n",
+				strerror(errno), at_least);
+			return map->base;
+		}
+
+		map->base = mmap(0, at_least, PROT_READ | PROT_WRITE,
+				 MAP_SHARED, map->fd, 0);
+		if (OFI_UNLIKELY(map->base == MAP_FAILED)) {
+			FI_WARN(&sm2_prov, FI_LOG_EP_CTRL,
+				"Failed to remap sm2_maps file when increasing "
+				"size to st.st_size=%ld, at_least, %zu, error: "
+				"%s\n",
+				st.st_size, at_least, strerror(errno));
+			abort();
```

Comment:
`sm2_mmap_remap()` is called from the fifo send/recv calls.  We don't have a clean path from the fifo send/recv calls to exit cleanly.  If remap fails during a send/recv, this is a critical error for us.  I changed all the other aborts to returns in this file b/c they were init only, but I think this is the special case b/c it is used during send/recv.

I am ok making this something other than an abort but I don't understand how we will cleanly communicate to the user that we have a critical problem.

Reply:
To make this more clear, the `sm2_mmap_remap` happens after `atomic_swap_ptr`, which changes the tail of the receivers fifo queue... If we cannot complete our transfer successfully, everything will hang, which means:
1. The receiver will be in a tight while loop waiting for us to complete our receive
2. All other senders sending to the receiver will push their Xfer Entries onto the receivers fifo queue (linked list) behind our partially completed send.

If this happens... the entire SHM communication on the instance is toast.

```
/* Write, Enqueue */
static inline void sm2_fifo_write(struct sm2_ep *ep, int peer_gid,
				  struct sm2_xfer_entry *xfer_entry)
{
...
	prev = atomic_swap_ptr(&peer_fifo->tail, offset);
	atomic_rmb();

	assert(prev != offset);

	if (OFI_LIKELY(SM2_FIFO_FREE != prev)) {
		if (OFI_UNLIKELY(prev + sizeof(xfer_entry) > map->size)) {
			/* Need to re-map */
			sm2_mmap_remap(map, prev + sizeof(xfer_entry));
			atomic_mb();
		}
...
```

Reply:
I understand that it's a critical issue but it would still be nice for the application to know about it. If there is a critical issue with the global shm file, then there's going to be a critical issue for the entire workload so aborting this one peer is not really going to do anything amazing. It would be more helpful for the application to be told this in some reasonable way so they can abort the entire job cleanly. Something like `return -FI_EAHHHHHHHHHH`
Maybe @shefty has opinions and insights (that's a joke because he always does)

Reply:
**NEVER** have a library call abort().  That's insane.  We're not the app.